### ISSUE TITLE
forcePage doesn't work

### ISSUE THREAD

**Issue by JakeElder** [AdeleD/react-paginate#124]
If I render this

```javascript
class MyPagination extends Component {
  constructor() {
    this.state = { currentPage: 0 }
  }

  render() {
    return (
      <ReactPaginate
        pageCount={10}
        pageRangeDisplayed={3}
        marginPageDisplayed={3}
        forcePage={this.state.currentPage}
      />
    )
  }
}
```

Then click on pages, the `selected` page changes. I would expect it to stay on page 0

I'd like to perform async actions after page click before having the change reflected. Is this possible given the current configuration options?

**Comment by jlarso11** [comment]
I just ran into this problem.  When sorting a list of objects, I want to go back to page 1 (index 0).  So I am using state to save that but it is no longer working.  It was working like a month ago

**Comment by vsuhasm** [comment]
I have the same problem. Trying to switch to the first page on user input in the search box.

**Comment by jlarso11** [comment]
I actually got this fixed with the new 4.1.1 version.  Make sure that you check the callback names.  They have changed

**Comment by jinxac** [comment]
Looks like this is still not working on 4.2.0. I will try to looking into it once. Will create a PR if i find a solution

**Comment by karthikiyengar** [comment]
Any update on what version this works with?

**Comment by nicoladj77** [comment]
I'm having the same issue too

**Comment by karthikiyengar** [comment]
@nicoladj77 - It does work for me actually, it was a mistake on my end. I had used the count of records directly as `pageCount` instead of providing the calculated value of pages. Maybe you have something similar?

**Comment by nicoladj77** [comment]
@karthikiyengar I've the following configuration

	render() {
		console.log(this.props);
		return <div className="m-pagination">
			<ReactPaginate previousLabel={"Previous"}
				nextLabel={"Next"}
				breakLabel={<span>...</span>}
				breakClassName={"page-numbers dots"}
				pageCount={this.props.pageCount}
				marginPagesDisplayed={2}
				pageRangeDisplayed={2}
				onPageChange={this.handlePageClick}
				forcePage={this.props.initialPage}
				initialPage={this.props.initialPage}
				containerClassName={"page numbers"}
				subContainerClassName={""}
				disabledClassName="disabled"
				hrefBuilder={this.hrefBuilder}
				activeClassName={"current"}
				disableInitialCallback={this.disableInitialCallback}/>
		</div>;
	}

My issue is the following:

A user navigates top page three. Then he changes the number of posts displayed from 10 to 20. I would want the component to reset itself to page one but it remains on page 3. pageCount is updated correctly and it's reflected in the component, but I can't force page 1.

**Comment by karthikiyengar** [comment]
@nicoladj77 - Are you setting the `forcePage` prop correctly after the number of posts have been changed? You are completely responsible for telling the component what page to set as the active page (even when you click next/prev or on a page directly) when you use `forcePage`. In your case, you would set the current page to 1 after you fetch data when the number of posts have changed.

**Comment by nicoladj77** [comment]
I'm setting initialPage ( which then it's used for forcePage ) to 0 because if I set it to 1 it sets page 2. the strange thing is that if I set it to 0, it doesn't change to page 1, but if I set it to 1, it does set the page as two. §It looks like the component is not updating. I'll look a little bit more into it

**Comment by nicoladj77** [comment]
Updated to version 4.3.1 and got it fixed somehow!

**Comment by karolis-inntec** [comment]
I'm using ^4.3.1, but still forcePage doesn't work.

**Comment by NourEldin275** [comment]
```forcePage``` works for me with version ```4.4.3```

**Comment by kevinch** [comment]
Any update on this? I still cannot get it to work on `4.4.3`. Second item keeps the `selected` class event though `forcePage` is set at 0.
@NourEldin275 could you tell us how you did it?

**Comment by NourEldin275** [comment]
@kevinch I simply do this 
```javascript
forcePage={this.state.currentPage}
```

And the `currentPage` gets updated with each page click. And when I want to, I set it to `0` and the first page gets selected. For example a reset button click handler or something like that.

Just make sure that `currentPage` is always set to `selected` value from the `onPageChange` handler that you pass to React Paginate like so:
```javascript
handlePageClick(data){
        let selected = data.selected;
        let offset = Math.ceil(selected * this.props.perPage);

        this.setState({offset: offset, currentPage: selected}, () => {
            //Get your data
        });
}
```

Note how I set the `currentPage: selected` in the handler you pass to React Paginate, this is to make sure that the current page is the one the user selected. And then when I need to set a specific page, you update the state object with the page you want and It'll re-render correctly.

Try this out and let me know if it works for you.

**Comment by kevinch** [comment]
I was messing up my `forcePage` via several `setState`s. Thanks for the answer! 

Btw, looks like this issue can be closed.

**Comment by dan-j** [comment]
This issue is still valid, and it seems to work perfectly if I change `componentWillReceiveProps` to not compare  `nextProps.forcePage` and always call
`this.setState({ selected: nextProps.forcePage })`

It's due to a race condition on `this.setState()`, on `handlePageSelected()`:

https://github.com/AdeleD/react-paginate/blob/d7a386befe32a7ebc615e698858c7b4dfd7470d3/react_components/PaginationBoxView.js#L88-L97

For example:

1. I am on page 0
2. click page 1.. this starts an async action, if I click back to page 0, `handlePageSelected` calls `setState({ selected: 0})`, my callback handler (using the `onPageChange` prop) ignores the request and causes my component to re-render with forcePage still equal to 1. 
4. At this point `componentWillReceiveProps()` is called, but `this.state.selected` is still 1 because the `setState()` on line 93 (point 2.) hasn't actually executed yet, so nothing is changed.
5. State is changed to 0, pagination component re-renders with 0 selected, even though forcePage=1

See below a screenshot of some logs explaining the flow above

![image](https://user-images.githubusercontent.com/5727701/30357240-fb892ac6-9880-11e7-861b-fd1f675370c0.png)

Logs starting `[Pagination]` is my component, `[react-paginate]` is this library's component. 

`12:29:17.829` is where it gets interesting, I select 0, my component says no, forcePage=1, componentDidReceiveProps ignores it, then it renders `{ selected: 0 }`

**Comment by dan-j** [comment]
To add to my previous comment, I think the best solution is inside `handlePageSelected()`, only call `this.setState()` if `this.props.forcePage` is not defined.

The other solution would to be use the callback on `this.setState(state, callback)` but this may cause multiple renders and potential flickering of pages between what the user clicked and what `forcePage` is set to.

**Comment by RuslanYurchenko** [comment]
Any progress on this? 
I have a several cases where I need to use forcePage. Cases with undefined by default work only once and then if state  doesn't change nothing happens.

**Comment by mlippens** [comment]
In my case this works perfectly by always setting `forcePage`, to the selected page when a page changes as suggested above. However, it seems that this does not fit all use cases people have.

**Comment by alvarohv** [comment]
Is this s a solution for this issue, I tried forcePage, but it is not working. 

I have the following code twice:

<ReactPaginate
                      previousLabel={'prev'}
                      nextLabel={'next'}
                      breakLabel={'...'}
                      breakClassName={'break-me'}
                      pageCount={ this.state.totalPages }
                      marginPagesDisplayed={2}
                      pageRangeDisplayed={ this.props.pageCount }
                      onPageChange={this.handlePageChange}
                      containerClassName={'pagination'}
                      pageClassName={"page-item"}
                      pageLinkClassName={"page-link"}
                      previousLinkClassName={"page-link prev"}
                      nextLinkClassName={"page-link next"}
                      disabledClassName={"disabled-page-nav"}
                      activeClassName={'active'}
                      forcePage={this.state.currentPage}
                    />

And I have this event to handle the click event:
  private handlePageChange = (event) => {
    this.setState({
      currentPage: event.selected
    }, () => { 
      this.setState({
        value: this.paginate(this.state.items)
      })
    });
  }

Any alternative to fix this issue?

**Comment by rflmyk** [comment]
Dude just do it:
```
  let forcePageObj = {}
  if (props.page === 0) {
    forcePageObj['forcePage'] = 0
  }

  return (
      <ReactPaginate
        onPageChange={handlePageChange}
        pageRangeDisplayed={3}
        marginPagesDisplayed={1}
        {...forcePageObj}
        pageCount={props.nbPages}
      />
  )
```

Like @sami616 did: https://github.com/AdeleD/react-paginate/issues/9#issuecomment-416545421

**Comment by marcomelilli** [comment]
@rflmyk it seems work only the first time. Then if I try to force to 0 another time it doesn't work...

**Comment by chiomadans1759** [comment]
> @karthikiyengar I've the following configuration
> 
> ```
> render() {
> 	console.log(this.props);
> 	return <div className="m-pagination">
> 		<ReactPaginate previousLabel={"Previous"}
> 			nextLabel={"Next"}
> 			breakLabel={<span>...</span>}
> 			breakClassName={"page-numbers dots"}
> 			pageCount={this.props.pageCount}
> 			marginPagesDisplayed={2}
> 			pageRangeDisplayed={2}
> 			onPageChange={this.handlePageClick}
> 			forcePage={this.props.initialPage}
> 			initialPage={this.props.initialPage}
> 			containerClassName={"page numbers"}
> 			subContainerClassName={""}
> 			disabledClassName="disabled"
> 			hrefBuilder={this.hrefBuilder}
> 			activeClassName={"current"}
> 			disableInitialCallback={this.disableInitialCallback}/>
> 	</div>;
> }
> ```
> 
> My issue is the following:
> 
> A user navigates top page three. Then he changes the number of posts displayed from 10 to 20. I would want the component to reset itself to page one but it remains on page 3. pageCount is updated correctly and it's reflected in the component, but I can't force page 1.

@nicoladj77, this is coming late but here is a solution.

The page number starts from 0 because the initial offset is supposed to be 0 (page * limit). However, for you to reset to page 1 when a user changes the number of posts displayed from 10 to 20, you have to reset the state page and offset to 0.

I utilized this in a search functionality where I needed to reset the page to 1 anytime I make a new request. So at the point of submitting the search input, I reset my page and offset to 0.

I hope this helps someone.

**Comment by chiomadans1759** [comment]
> Is this s a solution for this issue, I tried forcePage, but it is not working.
> 
> I have the following code twice:
> 
> <ReactPaginate
> previousLabel={'prev'}
> nextLabel={'next'}
> breakLabel={'...'}
> breakClassName={'break-me'}
> pageCount={ this.state.totalPages }
> marginPagesDisplayed={2}
> pageRangeDisplayed={ this.props.pageCount }
> onPageChange={this.handlePageChange}
> containerClassName={'pagination'}
> pageClassName={"page-item"}
> pageLinkClassName={"page-link"}
> previousLinkClassName={"page-link prev"}
> nextLinkClassName={"page-link next"}
> disabledClassName={"disabled-page-nav"}
> activeClassName={'active'}
> forcePage={this.state.currentPage}
> />
> 
> And I have this event to handle the click event:
> private handlePageChange = (event) => {
> this.setState({
> currentPage: event.selected
> }, () => {
> this.setState({
> value: this.paginate(this.state.items)
> })
> });
> }
> 
> Any alternative to fix this issue?

Here is a simple example. Anytime I select a page, I change my forcedPage to the selected page by updating the state value. And anytime I want to go back to page 1 when a method is called, I reset the page and offset back to 0
```
state = { 
      page: 0, 
      offset: 0,
      max: 10
}
```
```
                    <ReactPaginate
                        previousLabel={"previous"}
                        nextLabel={"next"}
                        breakLabel={<span>...</span>}
                        breakClassName={"break-me"}
                        pageCount={totalPages}
                        marginPagesDisplayed={2}
                        disableInitialCallback={true}
                        pageRangeDisplayed={5}
                        forcePage={this.state.page}
                        onPageChange={(page) => this.handlePageSelect(page)}
                        containerClassName={"pagination"}
                        subContainerClassName={"pages pagination"}
                        activeClassName={"active"}
                    />
```

```
handlePageSelect = (page) => {
        this.setState({
            filterParams: {
                ...this.state.filterParams,
                offset : page.selected * this.state.max,
                page: page.selected
            }
        },() => {
            this.searchMessages();
        });
    };
```

**Comment by chiomadans1759** [comment]
You need to use setState to change the page number and offset if applicable
back to 0 in the method you use to set number of posts displayed from 10 to
20. This should fix the issue

**Comment by hungvu193** [comment]
https://github.com/AdeleD/react-paginate/issues/198
This issue has been solved

**Comment by todd-drinkwater-dc** [comment]
I had an issue with using forcePage to control the state from the parent component where, like the above problems, it wasn't updating the state with the new page number.

In my case specifically, for my search parent component, I wanted it to reset the page number every time a search was done. The main issue is however, I set the state (that was passed down in the props) in the parent component as 0, instead of null. Because the only other time I was updating the props (and not the internal state), it was also setting the react-paginate props as 0 again, which led the component to think there was no need to re-render.

So in short.

- Set the parent components page number state as null to begin with, and pass this down to the react paginate component in the `forceState` prop.
- When you want to reset it again, update the state as usual in the parent component (if resetting the pagination to the start, still set the state to 0). Because 0 !== null, it will update the pagination component.
